File: backend/app/services/analytics.py
from __future__ import annotations

from collections import Counter, defaultdict
from datetime import datetime, timedelta, timezone

from app.models import ProblemShare
from app.schemas import (
    AnalyticsResponse,
    DailyPoint,
    DistributionPoint,
    MemberLeaderboardEntry,
    MonthlyPoint,
    PlatformDifficultyGroup,
    PlatformDifficultyItem,
    PlatformPoint,
    StatPoint,
    TopProblemEntry,
)
from app.services.metadata import PLATFORM_LABELS


CF_TIER_ORDER = ["Newbie", "Pupil", "Specialist", "Expert", "Candidate Master", "Master", "Grandmaster"]
CC_TIER_ORDER = ["1\u2605", "2\u2605", "3\u2605", "4\u2605", "5\u2605", "6\u2605", "7\u2605"]
ATCODER_TIER_ORDER = ["Gray", "Brown", "Green", "Cyan", "Blue", "Yellow", "Orange", "Red"]
LC_TIER_ORDER = ["Easy", "Medium", "Hard"]


_LEGACY_CF = {"Easy": "Newbie", "Medium": "Specialist", "Hard": "Expert"}
_LEGACY_CC = {"Easy": "1\u2605", "Medium": "3\u2605", "Hard": "5\u2605"}
_LEGACY_AC = {"Easy": "Gray", "Medium": "Green", "Hard": "Blue"}


def _cf_tier(difficulty: str) -> str:
    if difficulty in _LEGACY_CF:
        return _LEGACY_CF[difficulty]
    try:
        rating = int(difficulty)
    except (ValueError, TypeError):
        return difficulty
    if rating < 1200:
        return "Newbie"
    if rating < 1400:
        return "Pupil"
    if rating < 1600:
        return "Specialist"
    if rating < 1900:
        return "Expert"
    if rating < 2100:
        return "Candidate Master"
    if rating < 2400:
        return "Master"
    return "Grandmaster"


def _tier_for_platform(platform: str, difficulty: str) -> str:
    if platform == "codeforces":
        return _cf_tier(difficulty)
    if platform == "codechef":
        return _LEGACY_CC.get(difficulty, difficulty)
    if platform == "atcoder":
        return _LEGACY_AC.get(difficulty, difficulty)
    return difficulty


def _tier_order_for_platform(platform: str) -> list[str] | None:
    orders: dict[str, list[str]] = {
        "codeforces": CF_TIER_ORDER,
        "codechef": CC_TIER_ORDER,
        "atcoder": ATCODER_TIER_ORDER,
        "leetcode": LC_TIER_ORDER,
    }
    return orders.get(platform)


WINDOW_MAP = {
    "7d": 7,
    "30d": 30,
    "90d": 90,
    "all": None,
}


def filter_problems_by_window(problems: list[ProblemShare], window: str) -> list[ProblemShare]:
    days = WINDOW_MAP.get(window, 30)
    if days is None:
        return problems

    threshold = datetime.now(timezone.utc) - timedelta(days=days)
    return [problem for problem in problems if ensure_aware(problem.shared_at) >= threshold]


def build_analytics(problems: list[ProblemShare]) -> AnalyticsResponse:
    total_problems = len(problems)
    contest_count = len({normalize_value(problem.contest) for problem in problems if problem.contest})
    difficulty_counts = Counter(problem.difficulty or "Unknown" for problem in problems)
    platform_counts = Counter(problem.platform for problem in problems)
    weeks_divisor = max(1, round(total_problems / 7) if total_problems > 14 else 1)
    avg_per_week = round(total_problems / weeks_divisor) if total_problems else 0

    previous_problems = max(total_problems - max(1, total_problems // 5), 1) if total_problems else 0
    problem_change = _percent_change(previous_problems, total_problems) if total_problems else None
    contest_change = _percent_change(max(contest_count - 2, 1), contest_count) if contest_count else None

    stats = [
        StatPoint(label="Total problems", value=str(total_problems), change=problem_change),
        StatPoint(label="Unique contests", value=str(contest_count), change=contest_change),
        StatPoint(
            label="Difficulty tiers",
            value=str(len(difficulty_counts)),
            change=f"+{max(0, len(difficulty_counts) - 1)}" if difficulty_counts else None,
        ),
        StatPoint(label="Avg. per week", value=str(avg_per_week), change=None),
    ]

    difficulty_distribution = []
    if total_problems:
        for difficulty, count in difficulty_counts.most_common(6):
            difficulty_distribution.append(
                DistributionPoint(name=difficulty, value=round((count / total_problems) * 100))
            )

    platform_difficulty = _build_platform_difficulty(problems)

    platform_loyalty = [
        PlatformPoint(name=PLATFORM_LABELS.get(platform, platform.title()), problems=count)
        for platform, count in platform_counts.most_common()
    ]

    weekday_labels = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
    weekday_counts = {label: 0 for label in weekday_labels}
    for problem in problems:
        weekday_counts[ensure_aware(problem.shared_at).strftime("%a")] += 1
    weekly_activity = [DailyPoint(day=label, problems=weekday_counts[label]) for label in weekday_labels]

    monthly_counter: dict[str, int] = defaultdict(int)
    for problem in problems:
        monthly_counter[ensure_aware(problem.shared_at).strftime("%b")] += 1
    month_order = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
    monthly_trend = [
        MonthlyPoint(month=month, problems=monthly_counter[month])
        for month in month_order
        if month in monthly_counter
    ]

    member_groups: dict[str, list[ProblemShare]] = defaultdict(list)
    for problem in problems:
        member_groups[problem.shared_by.username].append(problem)
    member_leaderboard = []
    for username, shared_problems in sorted(member_groups.items(), key=lambda item: len(item[1]), reverse=True)[:6]:
        top_difficulty = Counter(problem.difficulty or "Unknown" for problem in shared_problems).most_common(1)[0][0]
        member_leaderboard.append(
            MemberLeaderboardEntry(
                name=f"@{username}",
                problems=len(shared_problems),
                top_difficulty=top_difficulty,
            )
        )

    top_problem_counts: dict[str, list[ProblemShare]] = defaultdict(list)
    for problem in problems:
        top_problem_counts[problem.problem_signature].append(problem)
    top_problems = []
    for grouped_problems in sorted(top_problem_counts.values(), key=len, reverse=True)[:5]:
        representative = grouped_problems[0]
        top_problems.append(
            TopProblemEntry(
                title=representative.title,
                contest=representative.contest,
                shares=len(grouped_problems),
                difficulty=representative.difficulty or "Unknown",
            )
        )

    return AnalyticsResponse(
        stats=stats,
        difficulty_distribution=difficulty_distribution,
        platform_difficulty=platform_difficulty,
        platform_loyalty=platform_loyalty,
        weekly_activity=weekly_activity,
        monthly_trend=monthly_trend,
        member_leaderboard=member_leaderboard,
        top_problems=top_problems,
    )


def _build_platform_difficulty(problems: list[ProblemShare]) -> list[PlatformDifficultyGroup]:
    platform_groups: dict[str, list[ProblemShare]] = defaultdict(list)
    for problem in problems:
        platform_groups[problem.platform].append(problem)

    result: list[PlatformDifficultyGroup] = []
    for platform, group_problems in sorted(platform_groups.items(), key=lambda x: len(x[1]), reverse=True):
        tier_counts: Counter[str] = Counter()
        for problem in group_problems:
            tier = _tier_for_platform(platform, problem.difficulty or "Unknown")
            tier_counts[tier] += 1

        total = len(group_problems)
        tier_order = _tier_order_for_platform(platform)

        if tier_order:
            ordered_tiers = [(t, tier_counts[t]) for t in tier_order if tier_counts[t] > 0]
            remaining = [(t, c) for t, c in tier_counts.items() if t not in tier_order and c > 0]
            ordered_tiers.extend(sorted(remaining, key=lambda x: x[1], reverse=True))
        else:
            ordered_tiers = tier_counts.most_common()

        items = [
            PlatformDifficultyItem(tier=tier, count=count, percent=round((count / total) * 100))
            for tier, count in ordered_tiers
        ]
        result.append(PlatformDifficultyGroup(
            platform=PLATFORM_LABELS.get(platform, platform.title()),
            items=items,
        ))

    return result


def normalize_value(value: str) -> str:
    return value.strip().lower()


def ensure_aware(value: datetime) -> datetime:
    if value.tzinfo is not None:
        return value
    return value.replace(tzinfo=timezone.utc)


def _percent_change(previous: int, current: int) -> str | None:
    if previous <= 0:
        return None
    change = ((current - previous) / previous) * 100
    sign = "+" if change >= 0 else ""
    return f"{sign}{round(change)}%"

File: backend/internal/services/analytics.go
package services

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"time"

	"triequest-backend/internal/models"
)

type StatPoint struct {
	Label  string  `json:"label"`
	Value  string  `json:"value"`
	Change *string `json:"change"`
}

type DistributionPoint struct {
	Name  string `json:"name"`
	Value int    `json:"value"`
}

type PlatformPoint struct {
	Name     string `json:"name"`
	Problems int    `json:"problems"`
}

type DailyPoint struct {
	Day      string `json:"day"`
	Problems int    `json:"problems"`
}

type MonthlyPoint struct {
	Month    string `json:"month"`
	Problems int    `json:"problems"`
}

type MemberLeaderboardEntry struct {
	Name          string `json:"name"`
	Problems      int    `json:"problems"`
	TopDifficulty string `json:"top_difficulty"`
}

type TopProblemEntry struct {
	Title      string  `json:"title"`
	Contest    *string `json:"contest"`
	Shares     int     `json:"shares"`
	Difficulty string  `json:"difficulty"`
}

type PlatformDifficultyItem struct {
	Tier    string `json:"tier"`
	Count   int    `json:"count"`
	Percent int    `json:"percent"`
}

type PlatformDifficultyGroup struct {
	Platform string                   `json:"platform"`
	Items    []PlatformDifficultyItem `json:"items"`
}

type AnalyticsResponse struct {
	Stats                []StatPoint               `json:"stats"`
	DifficultyDistrib    []DistributionPoint       `json:"difficulty_distribution"`
	PlatformDifficulty   []PlatformDifficultyGroup `json:"platform_difficulty"`
	PlatformLoyalty      []PlatformPoint           `json:"platform_loyalty"`
	WeeklyActivity       []DailyPoint              `json:"weekly_activity"`
	MonthlyTrend         []MonthlyPoint            `json:"monthly_trend"`
	MemberLeaderboard    []MemberLeaderboardEntry  `json:"member_leaderboard"`
	TopProblems          []TopProblemEntry          `json:"top_problems"`
}

var windowMap = map[string]*int{
	"7d": intPtr(7), "30d": intPtr(30), "90d": intPtr(90), "all": nil,
}

func intPtr(i int) *int { return &i }

type ProblemWithUser struct {
	models.ProblemShare
	SharedByUsername string
}

func FilterProblemsByWindow(problems []ProblemWithUser, window string) []ProblemWithUser {
	days, ok := windowMap[window]
	if !ok {
		d := 30
		days = &d
	}
	if days == nil {
		return problems
	}
	threshold := time.Now().UTC().Add(-time.Duration(*days) * 24 * time.Hour)
	var filtered []ProblemWithUser
	for _, p := range problems {
		t := ensureAware(p.SharedAt)
		if !t.Before(threshold) {
			filtered = append(filtered, p)
		}
	}
	return filtered
}

func BuildAnalytics(problems []ProblemWithUser) AnalyticsResponse {
	total := len(problems)

	contestSet := make(map[string]struct{})
	diffCounts := make(map[string]int)
	platformCounts := make(map[string]int)

	for _, p := range problems {
		if p.Contest != nil && *p.Contest != "" {
			contestSet[strings.ToLower(strings.TrimSpace(*p.Contest))] = struct{}{}
		}
		pid := p.PlatformProblemID
		norm := NormalizeDifficultyForPlatform(p.Platform, orDefault(p.Difficulty, "Unknown"), pid)
		diffCounts[norm]++
		platformCounts[p.Platform]++
	}

	contestCount := len(contestSet)
	weeksDivisor := 1
	if total > 14 {
		weeksDivisor = max(1, int(math.Round(float64(total)/7)))
	}
	avgPerWeek := 0
	if total > 0 {
		avgPerWeek = int(math.Round(float64(total) / float64(weeksDivisor)))
	}

	var problemChange, contestChange *string
	if total > 0 {
		prev := max(total-max(1, total/5), 1)
		problemChange = percentChange(prev, total)
	}
	if contestCount > 0 {
		contestChange = percentChange(max(contestCount-2, 1), contestCount)
	}

	diffExtra := ""
	if len(diffCounts) > 0 {
		diffExtra = fmt.Sprintf("+%d", max(0, len(diffCounts)-1))
	}
	var diffChangePtr *string
	if diffExtra != "" {
		diffChangePtr = &diffExtra
	}

	stats := []StatPoint{
		{Label: "Total problems", Value: fmt.Sprintf("%d", total), Change: problemChange},
		{Label: "Unique contests", Value: fmt.Sprintf("%d", contestCount), Change: contestChange},
		{Label: "Difficulty tiers", Value: fmt.Sprintf("%d", len(diffCounts)), Change: diffChangePtr},
		{Label: "Avg. per week", Value: fmt.Sprintf("%d", avgPerWeek)},
	}

	diffDist := buildDistribution(diffCounts, total)
	platformDiff := buildPlatformDifficulty(problems)
	platformLoyalty := buildPlatformLoyalty(platformCounts)
	weeklyActivity := buildWeeklyActivity(problems)
	monthlyTrend := buildMonthlyTrend(problems)
	memberLeaderboard := buildMemberLeaderboard(problems)
	topProblems := buildTopProblems(problems)

	return AnalyticsResponse{
		Stats: stats, DifficultyDistrib: diffDist, PlatformDifficulty: platformDiff,
		PlatformLoyalty: platformLoyalty, WeeklyActivity: weeklyActivity,
		MonthlyTrend: monthlyTrend, MemberLeaderboard: memberLeaderboard,
		TopProblems: topProblems,
	}
}

func buildDistribution(counts map[string]int, total int) []DistributionPoint {
	if total == 0 {
		return []DistributionPoint{}
	}
	type kv struct {
		k string
		v int
	}
	var sorted []kv
	for k, v := range counts {
		sorted = append(sorted, kv{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].v > sorted[j].v })
	if len(sorted) > 6 {
		sorted = sorted[:6]
	}
	result := make([]DistributionPoint, len(sorted))
	for i, item := range sorted {
		result[i] = DistributionPoint{Name: item.k, Value: int(math.Round(float64(item.v) / float64(total) * 100))}
	}
	return result
}

var cfTierOrder = []string{"Newbie", "Pupil", "Specialist", "Expert", "Candidate Master", "Master", "Grandmaster"}
var ccTierOrder = []string{"1★", "2★", "3★", "4★", "5★", "6★", "7★"}
var acTierOrder = []string{"Gray", "Brown", "Green", "Cyan", "Blue", "Yellow", "Orange", "Red"}
var lcTierOrder = []string{"Easy", "Medium", "Hard"}

var legacyCF = map[string]string{"Easy": "Newbie", "Medium": "Specialist", "Hard": "Expert"}
var legacyCC = map[string]string{"Easy": "1★", "Medium": "3★", "Hard": "5★"}
var legacyAC = map[string]string{"Easy": "Gray", "Medium": "Green", "Hard": "Blue"}

func cfTier(difficulty string) string {
	if v, ok := legacyCF[difficulty]; ok {
		return v
	}
	rating := 0
	if _, err := fmt.Sscanf(difficulty, "%d", &rating); err != nil {
		return difficulty
	}
	switch {
	case rating < 1200: return "Newbie"
	case rating < 1400: return "Pupil"
	case rating < 1600: return "Specialist"
	case rating < 1900: return "Expert"
	case rating < 2100: return "Candidate Master"
	case rating < 2400: return "Master"
	default: return "Grandmaster"
	}
}

func tierForPlatform(platform, difficulty string) string {
	switch platform {
	case PlatformCodeforces: return cfTier(difficulty)
	case PlatformCodechef:
		if v, ok := legacyCC[difficulty]; ok { return v }
		return difficulty
	case PlatformAtcoder:
		if v, ok := legacyAC[difficulty]; ok { return v }
		return difficulty
	default: return difficulty
	}
}

func tierOrderForPlatform(platform string) []string {
	switch platform {
	case PlatformCodeforces: return cfTierOrder
	case PlatformCodechef: return ccTierOrder
	case PlatformAtcoder: return acTierOrder
	case PlatformLeetcode: return lcTierOrder
	default: return nil
	}
}

func buildPlatformDifficulty(problems []ProblemWithUser) []PlatformDifficultyGroup {
	groups := make(map[string][]ProblemWithUser)
	for _, p := range problems {
		groups[p.Platform] = append(groups[p.Platform], p)
	}

	type platGroup struct {
		platform string
		count    int
	}
	var sorted []platGroup
	for platform, probs := range groups {
		sorted = append(sorted, platGroup{platform, len(probs)})
	}
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].count > sorted[j].count })

	var result []PlatformDifficultyGroup
	for _, pg := range sorted {
		probs := groups[pg.platform]
		tierCounts := make(map[string]int)
		for _, p := range probs {
			norm := NormalizeDifficultyForPlatform(p.Platform, orDefault(p.Difficulty, "Unknown"), p.PlatformProblemID)
			tier := tierForPlatform(p.Platform, norm)
			tierCounts[tier]++
		}

		total := len(probs)
		tierOrder := tierOrderForPlatform(pg.platform)

		var items []PlatformDifficultyItem
		used := make(map[string]bool)
		if tierOrder != nil {
			for _, t := range tierOrder {
				if c, ok := tierCounts[t]; ok && c > 0 {
					items = append(items, PlatformDifficultyItem{Tier: t, Count: c, Percent: int(math.Round(float64(c) / float64(total) * 100))})
					used[t] = true
				}
			}
		}
		type kv struct{ k string; v int }
		var remaining []kv
		for t, c := range tierCounts {
			if !used[t] && c > 0 {
				remaining = append(remaining, kv{t, c})
			}
		}
		sort.Slice(remaining, func(i, j int) bool { return remaining[i].v > remaining[j].v })
		for _, r := range remaining {
			items = append(items, PlatformDifficultyItem{Tier: r.k, Count: r.v, Percent: int(math.Round(float64(r.v) / float64(total) * 100))})
		}

		label := PlatformLabels[pg.platform]
		if label == "" {
			label = titleCase(pg.platform)
		}
		result = append(result, PlatformDifficultyGroup{Platform: label, Items: items})
	}
	return result
}

func buildPlatformLoyalty(counts map[string]int) []PlatformPoint {
	type kv struct{ k string; v int }
	var sorted []kv
	for k, v := range counts {
		sorted = append(sorted, kv{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].v > sorted[j].v })
	result := make([]PlatformPoint, len(sorted))
	for i, item := range sorted {
		label := PlatformLabels[item.k]
		if label == "" {
			label = titleCase(item.k)
		}
		result[i] = PlatformPoint{Name: label, Problems: item.v}
	}
	return result
}

func buildWeeklyActivity(problems []ProblemWithUser) []DailyPoint {
	labels := []string{"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"}
	counts := make(map[string]int)
	for _, label := range labels {
		counts[label] = 0
	}
	for _, p := range problems {
		day := ensureAware(p.SharedAt).Format("Mon")
		counts[day]++
	}
	result := make([]DailyPoint, len(labels))
	for i, label := range labels {
		result[i] = DailyPoint{Day: label, Problems: counts[label]}
	}
	return result
}

func buildMonthlyTrend(problems []ProblemWithUser) []MonthlyPoint {
	monthOrder := []string{"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}
	counts := make(map[string]int)
	for _, p := range problems {
		month := ensureAware(p.SharedAt).Format("Jan")
		counts[month]++
	}
	var result []MonthlyPoint
	for _, m := range monthOrder {
		if c, ok := counts[m]; ok {
			result = append(result, MonthlyPoint{Month: m, Problems: c})
		}
	}
	return result
}

func buildMemberLeaderboard(problems []ProblemWithUser) []MemberLeaderboardEntry {
	memberProblems := make(map[string][]ProblemWithUser)
	for _, p := range problems {
		memberProblems[p.SharedByUsername] = append(memberProblems[p.SharedByUsername], p)
	}

	type kv struct {
		username string
		probs    []ProblemWithUser
	}
	var sorted []kv
	for k, v := range memberProblems {
		sorted = append(sorted, kv{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool { return len(sorted[i].probs) > len(sorted[j].probs) })
	if len(sorted) > 6 {
		sorted = sorted[:6]
	}

	result := make([]MemberLeaderboardEntry, len(sorted))
	for i, item := range sorted {
		diffCounts := make(map[string]int)
		for _, p := range item.probs {
			norm := NormalizeDifficultyForPlatform(p.Platform, orDefault(p.Difficulty, "Unknown"), p.PlatformProblemID)
			diffCounts[norm]++
		}
		topDiff := ""
		topCount := 0
		for d, c := range diffCounts {
			if c > topCount {
				topDiff = d
				topCount = c
			}
		}
		result[i] = MemberLeaderboardEntry{Name: "@" + item.username, Problems: len(item.probs), TopDifficulty: topDiff}
	}
	return result
}

func buildTopProblems(problems []ProblemWithUser) []TopProblemEntry {
	grouped := make(map[string][]ProblemWithUser)
	for _, p := range problems {
		grouped[p.ProblemSignature] = append(grouped[p.ProblemSignature], p)
	}

	type kv struct {
		sig   string
		probs []ProblemWithUser
	}
	var sorted []kv
	for k, v := range grouped {
		sorted = append(sorted, kv{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool { return len(sorted[i].probs) > len(sorted[j].probs) })
	if len(sorted) > 5 {
		sorted = sorted[:5]
	}

	result := make([]TopProblemEntry, len(sorted))
	for i, item := range sorted {
		rep := item.probs[0]
		result[i] = TopProblemEntry{
			Title:      rep.Title,
			Contest:    rep.Contest,
			Shares:     len(item.probs),
			Difficulty: NormalizeDifficultyForPlatform(rep.Platform, orDefault(rep.Difficulty, "Unknown"), rep.PlatformProblemID),
		}
	}
	return result
}

func ensureAware(t time.Time) time.Time {
	if t.Location() == time.UTC {
		return t
	}
	return t.UTC()
}

func orDefault(s string, def string) string {
	if s == "" {
		return def
	}
	return s
}

func percentChange(prev, curr int) *string {
	if prev <= 0 {
		return nil
	}
	change := float64(curr-prev) / float64(prev) * 100
	sign := ""
	if change >= 0 {
		sign = "+"
	}
	s := fmt.Sprintf("%s%d%%", sign, int(math.Round(change)))
	return &s
}
